### PR TITLE
[IE CLDNN] Support IC={1,2,4} in the first convolution kernel

### PR DESCRIPTION
**Pull request by alex-chaiko**
This change extends first convolution kernel (bfyx_to_fsv16) to support cases with 1,2 or 4 channels instead of 3 as needed for some models.
Jira: CVS-33167

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/src/layout_optimizer.cpp [hunk 1]
```diff
@@ -209,13 +209,17 @@ bool layout_optimizer::can_fuse_reorder(program_node& prev, program_node& next,
...
-        (prev_output_layout.size.feature[0] == 3 || (prev_output_layout.size.feature[0] == 4 && (prev_dt == data_types::u8 || prev_dt == data_types::i8)))) ||
         (fmt_next == format::bs_fs_yx_bsv16_fsv16 && next_output_layout.size.feature[0] % 16 == 0 && prev_output_layout.size.feature[0] == 3) ||
         (fmt_next == format::bs_fs_yx_bsv16_fsv16 && next_output_layout.size.feature[0] >= 16 && prev_output_layout.size.feature[0] == 3 &&
         (next_output_layout.data_type != data_types::i8 && next_output_layout.data_type != data_types::u8))))
         return true;
 
+    if (next.is_type<convolution>() &&
+        fmt_prev == format::bfyx &&
+        fmt_next == format::b_fs_yx_fsv16 && next_output_layout.size.feature[0] >= 16 &&
+        (prev_output_layout.size.feature[0] <= 4 || (prev_output_layout.size.feature[0] == 4 && (prev_dt == data_types::u8 || prev_dt == data_types::i8))))
```
Comment (lznamens) [inline inference-engine/thirdparty/clDNN/src/layout_optimizer.cpp:220]: Looks like you can simplify this condition:
... (prev_output_layout.size.feature[0] <= 4))

File: inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_bfyx_to_b_fs_yx_fsv16.cpp [hunk 2]
```diff
@@ -169,6 +169,7 @@ JitConstants ConvolutionKernel_bfyx_to_bfyx_f16::GetJitConstants(const convoluti
     if (params.output.Feature().v % feature_block_size != 0) {
         jit.AddConstant(MakeJitConstant("OUTPUT_LEFTOVERS", 1));
     }
+    jit.AddConstant(MakeJitConstant("NUM_OF_CHANNELS", input.Feature().v));
```
[the remark this answers is not in the document]
Comment (alex-chaiko) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_kernel_bfyx_to_b_fs_yx_fsv16.cpp]: Changed to use INPUT0_FEATURE_NUM. I was not aware of this macro.

File: inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_to_bfyx_f16.cl [hunk 3]
```diff
@@ -133,27 +133,23 @@ KERNEL(convolution_bfyx_to_bfyx_f16)(
...
 
-                UNIT_TYPE src0 = intel_sub_group_shuffle(line_cache[0 * INPUT_BLOCK_SIZE + buf_offset], buf_group);
-                UNIT_TYPE src1 = intel_sub_group_shuffle(line_cache[1 * INPUT_BLOCK_SIZE + buf_offset], buf_group);
-                UNIT_TYPE src2 = intel_sub_group_shuffle(line_cache[2 * INPUT_BLOCK_SIZE + buf_offset], buf_group);
-
-                dst[i] = mad(wei[0], src0, dst[i]);
-                dst[i] = mad(wei[1], src1, dst[i]);
-                dst[i] = mad(wei2, src2, dst[i]);
+                UNIT_TYPE src[NUM_OF_CHANNELS];
+                for (int ic = 0; ic < NUM_OF_CHANNELS; ic++) {
```
[the remark this answers is not in the document]
Comment (alex-chaiko) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/convolution_gpu_bfyx_to_bfyx_f16.cl]: Looks like loop unrolling is happening, but I have added unroll hints to be sure. Although even with hints it is not guaranteed.

File: inference-engine/thirdparty/clDNN/src/layout_optimizer.cpp (hunk 1 above)
Comment (alex-chaiko) [reply inference-engine/thirdparty/clDNN/src/layout_optimizer.cpp]: Fixed.

**Comment by alex-chaiko** [comment]
@lznamens I added a change to address comment about block writes. Please take a look if it what you meant.

**Comment by alex-chaiko** [comment]
> #define UNIT_BLOCK_WRITEN(ptr, offset, val) CAT(UNIT_BLOCK_WRITE, OUTPUT_X_BLOCK_SIZE)(ptr, offset, val)

@lznamens I see, OUTPUT_X_BLOCK_SIZE is set to one of 2, 4, 8 in .cpp so that should work.

**Pull request merged**